extraction_history/Extraction_uhexp.re

```diff
@@ -0,0 +1,392 @@
+open Extraction_declear;
+open Extraction_trans;
+open Extraction_tool;
+open Extraction_uhpat;
+open Extraction_uhtyp;
+
+// This file will extract UHExp
+
+// directly use UHExp.string_of_operator to translate the operator
+//Js.log(UHExp.string_of_operator(Space));
+
+//TODO: add a new type STR, and modify pass_check.
+
+// FIXME: If we encounter "?" holes, for example in Case, if we leave it empty, it isn't None, but Some(Hole) indeed
+// It means when the user see a "? hole" but don't fill it, we receive Some(Hole) in AST. But if the user doesn't see a hole, it should be None
+// Currently, we don't have None cases except for enter ":" in "let"
+// TODO: It's safe and okey to transform Hole to None only for "? holes", i.e. call Extraction_tool.hole_to_none
+
+// The extract_t is prepared for ExpLine
+// In Letline, we can't actually evaluate the type for a let
+//      but we can evaluate the block in let
+//      Letline itself should have UNK, because we can't let forever
+let rec uhexp_trans = (~t: UHExp.t, ~vs: variable_set_t): extract_t =>
```

Comment:
these are just syntactic translations right? we had talked last semester about doing it in a bidirectional typing style. I think you should start with the code in Dynamics.Exp.syn_expand and modify it to output an OCaml string or OCaml AST instead of a DHExp, rather than writing something from scratch.